### PR TITLE
ThreadNet: add ShelleyAllegra tests

### LINKED ISSUE #2633: Include a ThreadNet test that hard forks from Shelley
We can start with forking from plain Shelley to plain Shelley. But forking to Multi-Asset Shelley would match the anticipated evolution of mainnet.

### PR DESCRIPTION
**Pull request by nfrisby**
Fixes #2633.

This follows up #2627 by adding tests that exercise the ability to hard from from a Shelley-based era to the next Shelley-based era (in this case, Shelley to Allegra).

This PR introduces the `TwoEras.hs` infrastructure. For every era we have, we should have a two-era test that has a primary interest of whether or not a proposal in that era can successfully induce a hard fork, especially in the presence of a short network partition. I suggest, as part of Issue #2472, we should continue to emphasize these "Will it fork?" tests.

As of this PR, `Cardano.hs` tests that we can hard fork out of Byron and `ShelleyAllegra.hs` tests that we can hard fork out of `Shelley`. At the moment, both have two eras, and so they can share the infrastructure that was factored out into `TwoEras.hs`

Eventually `Cardano.hs` will have 3-4 eras. At that time, I propose we introduce a `Byron______.hs` module that specifically tests leaving Byron. Maybe to Byron again, or still to Shelley, etc. To summarize: we should have a separate test dedicated to testing each adjacent pair of eras on mainnet and one more that tests a hard fork from the last mainnet era to another copy of itself.

Edit: Edsko and I have something along these lines in mind while working on Issue 2472, the rewrite. But we hope to have it be a tidy consolidated thing instead of as explicit as the above proposal.

### CODE REVIEW

**Comment by nfrisby** [comment]
This is timing out on the mac-mini build: `ShelleyShelley` tests give 900 seconds of silence. I'm going to add a debug tracers so that it's not 900 seconds of *silence* to see how long it actually takes.

**Comment by nfrisby** [comment]
1) I rebased, picking up diffs for most recent changes on master.

2) I added an extra commit that echos the milliseconds-since-process-start each time a ShelleyShelley invocation begins. This should prevent the "900 seconds of silence" on the mac-mini CI. Then we can inspect the logs and see how long it's actually taking. On the nightly runs, each ShelleyShelley invocation takes 40.5 seconds on average. And we're slashing the requested tests by 1/10, so I think this is running 10 tests, each of which is 40.5 seconds.

I'll admit, that's the first time I did the math. So if the variance is too high, maybe the total time is likely to be more than double the average... especially if the mac-mini build is relatively slow. I'll have some data after this CI build, and then we can consider disabling this test outside of the nightly.

**Comment by nfrisby** [comment]
My latest push here disables random transactions in `ShelleyShelley`. That gives a 10x speed-up. (See input-output-hk/cardano-ledger-specs#1901.)

Remaining item is to fix the translation of the ledger view forecast.

**Comment by mrBliss** [comment]
I'll take over this PR so that @nfrisby can focus on the ThreadNet rewrite.

> My latest push here disables random transactions in `ShelleyShelley`. That gives a 10x speed-up. (See [input-output-hk/cardano-ledger-specs#1901](https://github.com/input-output-hk/cardano-ledger-specs/issues/1901).)
> 

I'll rebase this tomorrow, not that #2681 has been merged. That PR disables the Shelley transaction generator, which this PR also does.

> Remaining item is to fix the translation of the ledger view forecast.

The master branch includes #2661, which provides a correct translation function which I'll use.


> we should have a separate test dedicated to testing each adjacent pair of eras on mainnet

Agreed. As of #2666, we have the Allegra and Mary eras. So I'm thinking of changing `ShelleyShelley` to `ShelleyAllegra`.

At the moment, there is no real difference between Shelley->Shelley, Shelley->Allegra, or Allegra->Mary (Allegra and Mary are currently identical to Shelley). That will soon change with #2679, after which I'll also introduce `MaryAllegra`. I think the pattern synonyms from `ShelleyShelley` can be generalised to any two eras.

I would still like to have at least *one* test running *all* the hard forks of `CardanoBlock`, i.e., all 3 of them (Byron->Shelley->Allegra->Mary), since that's what mainnet is running.

**Comment by nfrisby** [comment]
> I would still like to have at least one test running all the hard forks of CardanoBlock, i.e., all 3 of them (Byron->Shelley->Allegra->Mary), since that's what mainnet is running.

Yep, that's what I had in mind 👍 

Do note: such a test will require a lot of slots; those epochs are long. Also, the current ThreadNet tests cannot easily (at all?) express the necessary transactions for transitioning from eg the second era to the third. At the moment, the "crucial tx" infra we use for these can only (easily?) express the transactions necessary for leaving the first era. (This is compounded by needing to know a valid UTxO for use as an input.)

Fixing this is something we're keeping in mind for the ThreadNet rewrite.

Thanks for taking this over.

**Comment by mrBliss** [comment]
FYI, I have been working on this in the background. I have refactored some things to make the two-era infrastructure generic and reusable with a minimum of boilerplate (e.g., #2687). I am currently distracted by some other ledger integration related things, but will soon get back to this.

**Review by mrBliss — COMMENTED** [review]
Update:
* Rebased
* Removed some unused code from `TwoEras` that was copied from `Cardano`, e.g., `TestSetup`.
* Changed the names in `TwoEras` to no longer assume Byron and Shelley. See the TODOs.
* Generalised the `ShelleyShelley` infra to `ShelleyBasedHardFork`. This generalisation ironically does *not* permit `ShelleyShelley` as it wants a translation between the two eras.
* Changed the `ShelleyShelley` test to `ShelleyAllegra`, as this is the next hard fork, so it's important to test it. (Note that we will extend `Cardano` with Allegra and Mary soon).

I have left some comments on the TODOs that still need addressing.

File: ouroboros-consensus-cardano-test/test/Test/ThreadNet/ShelleyAllegra.hs [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+                        shelleyGenesis           = genesisShelley
+                      , shelleyInitialNonce      = setupInitialNonce
+                      , shelleyProtVer           = SL.ProtVer majorVersion1 0
+                      , shelleyLeaderCredentials = Identity $
+                          Shelley.mkLeaderCredentials (coreNodes !! fromIntegral nid)
+                      }
+                    ProtocolParamsAllegra {
+                        allegraProtVer           = SL.ProtVer majorVersion2 0
+                      , allegraLeaderCredentials = Identity $
+                          -- TODO different credentials?
```
[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus-cardano-test/test/Test/ThreadNet/ShelleyAllegra.hs:278]: This is fine, as we're using MockKES anyway.

To handle it for a real node, we could consider sharing a `HotKey` across eras.

File: ouroboros-consensus-cardano-test/test/Test/ThreadNet/ShelleyAllegra.hs [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+          , (blk, _m) <- maybeToList $ Map.maxView nodeOutputForges
+                -- the last block the node forged
+          ]
+      , rsEra2Slots  =
+          --- TODO this comment and code are wrong
+
+          BoolProps.requiredIf $
+          -- The active slots in the first two Shelley epochs are all overlay
+          -- slots, so the first Shelley block will arise from one of those.
+          not $ Set.null overlaySlots
```
[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus-cardano-test/test/Test/ThreadNet/ShelleyAllegra.hs:345]: @nfrisby I don't understand this. Should this be fixed before merging? I'm inclined to leave it as-is, as the rewrite is coming.

**Review by edsko — APPROVED** [review]

File: ouroboros-consensus-cardano-test/src/Test/ThreadNet/Infra/ShelleyBasedHardFork.hs [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+pattern GenTxShelley2 tx = HardForkGenTx (OneEraGenTx (S (Z tx)))
+
+{-# COMPLETE GenTxShelley1, GenTxShelley2 #-}
+
+pattern ShelleyBasedHardForkNodeToNodeVersion1 ::
+     BlockNodeToNodeVersion (ShelleyBasedHardForkBlock era1 era2)
+pattern ShelleyBasedHardForkNodeToNodeVersion1 =
+    HardForkNodeToNodeEnabled
+      HardForkSpecificNodeToNodeVersion1
+      (  EraNodeToNodeEnabled ShelleyNodeToNodeVersion1
```
[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus-cardano-test/src/Test/ThreadNet/Infra/ShelleyBasedHardFork.hs:98]: Uggh, can't use `maxBound` here, this also has to be a pattern of course.

File: ouroboros-consensus-test/src/Test/ThreadNet/TxGen.hs [hunk 4]
```diff
@@ -39,3 +50,58 @@ class TxGen blk where
...
+--
+-- We don't provide a generic instance of 'TxGen' because it might be desirable
+-- to provide custom implementations for specific instantiations of the eras of
+-- 'HardForkBlock'. Instead, we provide this function that can be used when a
+-- generic implemenation is desired.
+--
+-- Choose @NP WrapTxGenExtra xs@ for the instance of the 'TxGenExtra' type
+-- family, where @xs@ matches the concrete instantiation.
+testGenTxsHfc ::
+     forall xs. (All TxGen xs, CanHardFork xs)
```
[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus-test/src/Test/ThreadNet/TxGen.hs:74]: That would require taking an NP of `testGenTxs`, which is annoying. Leave it.

File: ouroboros-consensus-cardano-test/src/Test/ThreadNet/TxGen/Allegra.hs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+{-# OPTIONS_GHC -Wno-orphans #-}
+module Test.ThreadNet.TxGen.Allegra () where
+
+import           Ouroboros.Consensus.Shelley.Eras
+import           Ouroboros.Consensus.Shelley.Ledger
+
+import           Test.ThreadNet.TxGen (TxGen (..))
+
+-- | Dummy generator until CAD-2119 is done, i.e., the transaction generator in
+-- the ledger has been generalised over the eras.
```
[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus-cardano-test/src/Test/ThreadNet/TxGen/Allegra.hs:12]: https://github.com/input-output-hk/cardano-ledger-specs/pull/1997 has been merged, but there isn't yet an instance for Allegra/Mary as far as I can see. https://github.com/input-output-hk/cardano-ledger-specs/pull/2000 is open.

**Comment by mrBliss** [comment]
Since the ThreadNet tests pass (including a nightly) with some of the Byron-specific bits in place, I'm going to merge it as-is. We can improve them later.

**Comment by mrBliss** [comment]
bors merge

**Comment by mrBliss** [comment]
bors merge

**Comment by mrBliss** [comment]
bors ping

**Comment by mrBliss** [comment]
bors merge

**Comment by mrBliss** [comment]
Bors' job in Hydra finished 1h ago: https://hydra.iohk.io/eval/981006 :roll_eyes:

**Comment by mrBliss** [comment]
bors merge

**Pull request merged**